### DISCUSSION TITLE
Integrate SingleFIle into NodeJS

### DISCUSSION THREAD

**Discussion by daniel31x13** [gildas-lormeau/single-file-cli#73] [General]
Hello,
I'm currently working on an [app](https://github.com/Daniel31x13/link-warden) and wanted to know how can I run SingleFile from NodeJS.

Cheers,
Daniel

**Comment by gildas-lormeau** [comment]
Hi Daniel,

I'm glad to hear that you would like to integrate SingleFile in your app :). There are 2 main ways to run SingleFile from NodeJS. 
You can use JSDOM, see https://github.com/gildas-lormeau/single-file-cli/blob/master/back-ends/jsdom.js, and run SingleFile directly in Node. In this case, the JavaScript won't be executed and the resulting page might be altered. For example, lazy-loaded images might be missing.
Otherwise, you can run SingleFile in a headless browser via puppeteer for example, see other files here: https://github.com/gildas-lormeau/single-file-cli/tree/master/back-ends. I do not recommend to use Selenium though. With puppeteer, you should get more or less the same result as the extension.
An overview of the integration of SingleFile is explained here: https://github.com/gildas-lormeau/SingleFile/wiki/How-to-integrate-SingleFile-library-code-in-%22custom%22-environments%3F

Please let me know if you need more details.

**Comment by gildas-lormeau** [comment]
I'm closing and moving this issue in the "Discussions" tab.

**Comment by bryanph** [comment]
Having tried both JSDOM and puppeteer, my suggestion would be to avoid JSDOM. In my experience It's too slow for doing anything more than basic tasks. Seeing as SingleFile does a lot of parsing of content, it will probably be way too slow.

**Comment by CRC32EX** [comment]
Puppeteer supports `headful mode`
So, we can integrate chrome extensions easily, I think.

Refs
- https://twitter.com/ChromiumDev/status/1628439484043001856
- https://pptr.dev/guides/chrome-extensions

**Reply by gildas-lormeau** [reply]
Thank you for the info! That's very interesting :)

**Comment by daniel31x13** [comment]
@gildas-lormeau Hey there it’s been a while, sorry for not responding.
Long story short, I did a full rebuild since the initial version lacked many features from the ground up.
So anyway everything has been changed/improved…

Coming back to this, what is the best way you’d suggest me to implement SingleFile into Linkwarden? Worth noting that I’m using Playwright behind the scenes.

The closest thing I found is https://github.com/gildas-lormeau/SingleFile/issues/723#issuecomment-885288275 which didn't work out for me.

**Reply by gildas-lormeau** [reply]
@daniel31x13 Hi, I just ported the CLI tool to Deno in order to produce executable files easily (see https://github.com/gildas-lormeau/single-file-cli/releases). It now relies directly on the Chrome DevTools Protocol to pilot the browser. If you want to use Playwright, then you could take inspiration from this script maybe: https://github.com/gildas-lormeau/single-file-cli/blob/0cb134df0638f8c7f198097d8ff830a16f8b9f9d/back-ends/playwright-chromium.js. Otherwise you could try to port https://github.com/gildas-lormeau/single-file-cli/blob/5b9d11bc0b1a9dfe08d92207045c9d6d09c24fbf/lib/cdp-client.js to make it compatible with Playwright.

**Reply by daniel31x13** [reply]
@gildas-lormeau Thanks! Gonna figure it out.

Question, is the final html file XSS-proof? Is `single-file` stripping out the JS parts that might show vulnerability when posted online?

**Reply by gildas-lormeau** [reply]
I confirm that by default all the scripts are removed from the page. The [content security policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) added or updated by SingleFile in the saved page is also making sure that all accesses to external resources are blocked.

**Reply by daniel31x13** [reply]
@gildas-lormeau Hey sorry for keep bothering.

So [here's](https://github.com/linkwarden/linkwarden/blob/2b8f7d4be20026705fc99db18d9f38c7c0f8a7d3/lib/api/preservationScheme/archiveAsSinglefile.ts#L16) how we'll be implementing it, basically we have an environment variable called `SINGLEFILE_ARCHIVE_COMMAND` and we'll be replacing the `{{URL}}` with the actual link that we want to archive:
```
SINGLEFILE_ARCHIVE_COMMAND=deno run --allow-run --allow-write --allow-net --allow-read /path/to/single-file {{URL}} --dump-content
```
---
Now the main issue is that [some links](https://liftvault.com/programs/strength/) do not exit successfully when using `execSync`, causing it to wait indefinitely. With a 2-minute timeout, the process fails to archive correctly.

---

### TL;DR

If you run the following command you'll notice it generates the html output, but it doesn't get exited properly. I need it to be exited so that I can get the output using `execSync`.

```
deno run --allow-run --allow-write --allow-net --allow-read /path/to/single-file [{{URL}} ](https://liftvault.com/programs/strength/) --dump-content
```

![image](https://github.com/gildas-lormeau/SingleFile/assets/93432314/3342126b-573f-462f-b130-20776e078cfc)

**Reply by gildas-lormeau** [reply]
I cannot reproduce the problem on macOS. If I run the command below from the command line interface, the program exits as expected. On what OS did you do your test?

```sh
deno run --allow-run --allow-write --allow-net --allow-read ./single-file https://liftvault.com/programs/strength/ --dump-content
```

**Reply by daniel31x13** [reply]
Strange, I'm also on mac. Gonna test it on a VPS instance and see if it works. Maybe it has something to do with my local configurations...

**Reply by daniel31x13** [reply]
Ok so I went for the [manual installation](https://github.com/gildas-lormeau/single-file-cli?tab=readme-ov-file#manual-installation) on a dummy server, installed `chromium-browser` using apt and now it doesn't work at all. Just stuck on a blinking cursor state:
![image](https://github.com/gildas-lormeau/SingleFile/assets/93432314/b307d573-c3fd-4adf-b7af-56f027751aee)

By any chance do you know what has possibly gone wrong? Is running `apt install chromium-browser` enough for it to work?
Per the [intro](https://github.com/gildas-lormeau/single-file-cli?tab=readme-ov-file#introduction):

> It runs through Deno as a standalone script injected into the web page via the Chrome DevTools Protocol instead of being embedded into a WebExtension.

**Reply by gildas-lormeau** [reply]
I did a test in Fedora and it works fine for me. Could you try to update single-file-cli and run your test again? I fixed some issues related to error handling which could help to understand the cause of the issue.

**Reply by daniel31x13** [reply]
Just created a fresh server, went the exact steps and the exact error appeared. (hanging like above...)

I'm guessing the problem is because installing `chromium-browser` isn't enough and there is another dependancy that's required for servers, also it's worth mentioning that the server is Ubuntu.

**Reply by gildas-lormeau** [reply]
FYI, I did a test with Ubuntu Server 22.04.4 (64-bit ARM) in a VM (with the user name "ubuntu", 4GB of RAM and 32GB of disk) downloaded from https://cdimage.ubuntu.com/releases/jammy/release/. It worked as expected on my end. I ran these commands:

```sh
sudo apt install chromium-browser
sudo apt install npm
sudo apt install unzip
curl -fsSL https://deno.land/install.sh | sh
export DENO_INSTALL="/home/ubuntu/.deno"
export PATH="$DENO_INSTALL/bin:$PATH"
npm install single-file-cli
cd node-modules/single-file-cli/
deno run --allow-run --allow-write --allow-net --allow-read ./single-file https://liftvault.com/programs/strength/ --dump-content
```

**Reply by daniel31x13** [reply]
Hmm, this is pretty strange since I'm following the _exact_ steps and encountering the same issue. The config of the dummy server is also pretty much the same as yours (with the exception that it's not ARM-based... Idk if that really matters.)

Email me your SSH public key and I can give you access to the dummy server to take a look _if you're okay with that_... It's a fresh server I created a couple of minutes ago.

Also thanks for putting your time till now!

**Reply by gildas-lormeau** [reply]
Sorry for the late reply, I fixed some issues meanwhile that could be the cause of the problem. Can you test the last version please? If it still does not work, I'll give you my public key.

**Reply by daniel31x13** [reply]
Unfortunately the exact same issue is persisting.

**Reply by gildas-lormeau** [reply]
Sorry for the late reply, do you still need help?
